File: src/main/javabak/materials/Material.java
package redcrafter07.processed.materials;

import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.ItemTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour;

import javax.annotation.Nullable;

public class Material {
    private final String identifier;
    private final int color;
    private final String chemicalDescription;
    private final @Nullable TagKey<Block> materialTag;
    private final BlockBehaviour.Properties metalBlockProperties;
    private final BlockBehaviour.Properties oreProperties;

    public Material(String identifier,
                    int color,
                    String chemicalDescription,
                    @Nullable TagKey<Block> materialTag,
                    BlockBehaviour.Properties metalBlockProperties,
                    BlockBehaviour.Properties oreProperties) {
        this.identifier = identifier;
        this.color = color;
        this.chemicalDescription = chemicalDescription;
        this.materialTag = materialTag;
        this.metalBlockProperties = metalBlockProperties;
        this.oreProperties = oreProperties;
    }

    public MutableComponent getComponent() {
        return Component.translatable("processed.material." + identifier);
    }

    public int color() {
        return color;
    }

    public String chemicalDescription() {
        return chemicalDescription;
    }

    public String getDustPath() {
        return identifier + "_dust";
    }

    public TagKey<Item> getDustTag() {
        return ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "dusts/" + identifier));
    }

    public String getIngotPath() {
        return identifier + "_ingot";
    }

    public TagKey<Item> getIngotTag() {
        return ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "ingots/" + identifier));
    }

    public String getNuggetPath() {
        return identifier + "_nugget";
    }

    public TagKey<Item> getNuggetTag() {
        return ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "nuggets/" + identifier));
    }

    public String getRawPath() {
        return "raw_" + identifier;
    }

    public TagKey<Item> getRawTag() {
        return ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "raw_materials/" + identifier));
    }

    public String getMetalBlockPath() {
        return identifier + "_block";
    }

    public BlockBehaviour.Properties getMetalBlockProperties() {
        return metalBlockProperties;
    }

    public String getOreBlockPath() {
        return identifier + "_ore";
    }

    public BlockBehaviour.Properties getOreBlockProperties() {
        return oreProperties;
    }

    public @Nullable TagKey<Block> getMaterialTag() {
        return materialTag;
    }
}

File: src/main/kotlin/redcrafter07/processed/materials/Material.kt
package redcrafter07.processed.materials

import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.chat.MutableComponent
import net.minecraft.resources.ResourceLocation
import net.minecraft.tags.ItemTags
import net.minecraft.tags.TagKey
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockBehaviour
import redcrafter07.processed.Translations
import redcrafter07.processed.rl

open class Material(
    val info: MaterialInfo,
    val materialTag: TagKey<Block>,
    val metalBlockProperties: BlockBehaviour.Properties,
    val oreBlockProperties: BlockBehaviour.Properties
) {
    val component: MutableComponent get() = Translations.materialName(identifier)

    val identifier: String get() = info.identifier
    val dustPath: String get() = "${identifier}_dust"
    val rawPath: String get() = "raw_${identifier}"
    val nuggetPath: String get() = "${identifier}_nugget"
    val ingotPath: String get() = "${identifier}_ingot"
    val metalBlockPath: String get() = "${identifier}_block"
    val rawMetalBlockPath: String get() = "raw_${identifier}_block"
    val oreBlockPath: String get() = "${identifier}_ore"

    val dust: Item get() = BuiltInRegistries.ITEM.get(rl(dustPath))
    val raw: Item get() = BuiltInRegistries.ITEM.get(rl(rawPath))
    val nugget: Item get() = BuiltInRegistries.ITEM.get(rl(nuggetPath))
    val ingot: Item get() = BuiltInRegistries.ITEM.get(rl(ingotPath))

    val metalBlockItem: Item get() = BuiltInRegistries.ITEM.get(rl(metalBlockPath))
    val rawMetalBlockItem: Item get() = BuiltInRegistries.ITEM.get(rl(rawMetalBlockPath))
    val oreBlockItem: Item get() = BuiltInRegistries.ITEM.get(rl(oreBlockPath))

    val metalBlock: Block get() = BuiltInRegistries.BLOCK.get(rl(metalBlockPath))
    val rawMetalBlock: Block get() = BuiltInRegistries.BLOCK.get(rl(rawMetalBlockPath))
    val oreBlock: Block get() = BuiltInRegistries.BLOCK.get(rl(oreBlockPath))

    val dustTag: TagKey<Item> get() = ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "dusts/$identifier"))
    val rawTag: TagKey<Item> get() = ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "raw_materials/$identifier"))
    val nuggetTag: TagKey<Item> get() = ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "nuggets/$identifier"))
    val ingotTag: TagKey<Item> get() = ItemTags.create(ResourceLocation.fromNamespaceAndPath("c", "ingots/$identifier"))

    fun <T> getExtraData(clazz: Class<T>): T? = info.getExtraData(clazz)

    override fun toString(): String = "Material[$identifier]"
}